planner/core/rule_predicate_push_down.go

```diff
@@ -461,7 +496,8 @@ func (p *LogicalJoin) outerJoinPropConst(predicates []expression.Expression) []e
 	p.LeftConditions = nil
 	p.RightConditions = nil
 	p.OtherConditions = nil
-	joinConds, predicates = expression.PropConstOverOuterJoin(p.ctx, joinConds, predicates, outerTable.Schema(), innerTable.Schema())
+	isAntiJoin := p.JoinType == AntiLeftOuterSemiJoin
```

Comment:
should we consider `LeftOuterSemiJoin`/`AntiSemiJoin`/`SemiJoin`?

Reply:
From the comment of https://github.com/pingcap/tidb/pull/8706#issuecomment-449952418, I think we should not generate a `NOT NULL` filter for the inner join keys on `LeftOuterSemiJoin`. 

Take the following case as an example:
```sql
drop table if exists t;
create table t (a bigint, b bigint);
insert into t values(null, 1), (2, 2), (3, null);
```

In MySQL:
```sql
MySQL(root@localhost:test) > select a, a in (select b from t) from t;
+------+------------------------+
| a    | a in (select b from t) |
+------+------------------------+
| NULL |                   NULL |
|    2 |                      1 |
|    3 |                   NULL |
+------+------------------------+
3 rows in set (0.00 sec)
```

In TiDB:
```sql
TiDB(root@127.0.0.1:test) > desc select a, a in (select b from t) from t;
+-----------------------+-------+------+----------------------------------------------------------------------------+
| id                    | count | task | operator info                                                              |
+-----------------------+-------+------+----------------------------------------------------------------------------+
| HashLeftJoin_7        | 3.00  | root | left outer semi join, inner:TableReader_11, equal:[eq(test.t.a, test.t.b)] |
| ├─TableReader_9       | 3.00  | root | data:TableScan_8                                                           |
| │ └─TableScan_8       | 3.00  | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                 |
| └─TableReader_11      | 3.00  | root | data:TableScan_10                                                          |
|   └─TableScan_10      | 3.00  | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                 |
+-----------------------+-------+------+----------------------------------------------------------------------------+
5 rows in set (0.00 sec)

TiDB(root@127.0.0.1:test) > select a, a in (select b from t) from t;
+------+------------------------+
| a    | a in (select b from t) |
+------+------------------------+
| NULL |                      0 |
|    3 |                      0 |
|    2 |                      1 |
+------+------------------------+
3 rows in set (0.00 sec)
```

TiDB produces wrong result in this case:
1. `NULL 0` should be `NULL NULL`, because the outer join key is `NULL`
2. `3 0` should be `3 NULL`, because there is no inner join key valued `3`, and there is a join key which is `NULL`

So if we generates a `NOT NULL` filter on the inner join key in this case, we'll never know that there is a join key in the inner side. Thus we can not produce the correct result for outer join key `3`.